### ISSUE TITLE
espanso running but not expanding

### ISSUE THREAD

**Issue by maxandersen** [federico-terzi/espanso#216]
not sure what triggers it but i noticed multiple times expansion stops working and now its just fully stopped.

i'm on OSX 10.15.3

```
❯ uname -a
Darwin darkbeard.local 19.3.0 Darwin Kernel Version 19.3.0: Thu Jan  9 20:58:23 PST 2020; root:xnu-6153.81.5~1/RELEASE_X86_64 x86_64
❯ espanso status
espanso is running
❯ espanso log
05:38:41 [ INFO] espanso version 0.5.1
05:38:41 [ INFO] using config path: /Users/max/Library/Preferences/espanso
05:38:41 [ INFO] using package path: /Users/max/Library/Application Support/espanso/packages
05:38:41 [ INFO] starting daemon...
05:38:41 [ INFO] Status icon already initialized, skipping.
05:38:41 [ INFO] Initializing EspansoNotifyHelper in /Users/max/Library/Application Support/espanso
05:38:41 [ INFO] EspansoNotifyHelper already initialized, skipping.
05:38:41 [ INFO] Binded to IPC unix socket: /Users/max/Library/Application Support/espanso/espanso.sock
05:38:41 [ INFO] espanso is running!
```

i've tried espanso restart and it seem to make no difference.

**Comment by federico-terzi** [comment]
Hey @maxandersen,

I'm sorry you are experiencing these issues, I still haven't figured out why, but after the Catalina update things got problematic on macOS. I have a couple of questions for you:

* Does it always happen or only after waking up from sleep?

In one of the moments in which espanso becomes unresponsive, could you please execute these commands and post the output here?


```
ioreg -l -w 0 | grep SecureInput
```

> That will tell you the process ID (kCGSSessionSecureInputPID) of the application that has Secure Input enabled. 

Could you also post the output of the following command in case the previous one returns something? (replacing NNN with the process ID found before)

```
ps auxww | grep NNN
```



Cheers :)

**Comment by maxandersen** [comment]
I haven't spotted the pattern yet but from sleep could be the case ...but not sure.

```
ioreg -l -w 0 | grep SecureInput | grep kCGSSessionSecure
  |   "IOConsoleUsers" = ({"kCGSSessionOnConsoleKey"=Yes,"kSCSecuritySessionID"=100006,"kCGSSessionSecureInputPID"=64936,"kCGSSessionGroupIDKey"=20,"kCGSessionLoginDoneKey"=Yes,"kCGSSessionSystemSafeBoot"=No,"kCGSSessionUserNameKey"="max","kCGSSessionIDKey"=257,"kCGSessionLongUserNameKey"="Max Rydahl Andersen","kCGSSessionAuditIDKey"=100006,"kCGSSessionLoginwindowSafeLogin"=No,"kCGSSessionUserIDKey"=501})
    | |   "IOConsoleUsers" = ({"kCGSSessionOnConsoleKey"=Yes,"kSCSecuritySessionID"=100006,"kCGSSessionSecureInputPID"=64936,"kCGSSessionGroupIDKey"=20,"kCGSessionLoginDoneKey"=Yes,"kCGSSessionSystemSafeBoot"=No,"kCGSSessionUserNameKey"="max","kCGSSessionIDKey"=257,"kCGSessionLongUserNameKey"="Max Rydahl Andersen","kCGSSessionAuditIDKey"=100006,"kCGSSessionLoginwindowSafeLogin"=No,"kCGSSessionUserIDKey"=501})
```

```
ps auxww | grep 64936
max              64936  11.0  0.3  6681068 108220   ??  S     4:27PM   0:06.83 /Applications/iTerm.app/Contents/MacOS/iTerm2
```
 
so the id points to iterm 2...if I restart iterm2 then  espanso works again!

Does that also explain why I can't use espanso in iterm2 in general ?

**Comment by federico-terzi** [comment]
Hey @maxandersen,

I think that could be it! Searching online it seems that iTerm2 enables SecureInput by default (which makes the OS block any text expansion). Luckily, it seems like you can disable SecureInput on iTerm2 quite easily, check out [this forum post](https://gitlab.com/gnachman/iterm2/issues/3937).

In that way, you should be able to use espanso without problems. Let me know if that works as expected and if you have other doubts.

Actually your answer is pretty useful for the project in general, as I'll now find a way to automatically notify the user in these situations, so that they know why espanso is not working.

Thanks for your help :)

**Comment by maxandersen** [comment]
yeah, I noticed textexpander actually had a popup warning about process X by name having blocked secure input. would be nice with similar warning.

I'll try disable iterm2 secureinput and see if that helps me ;)

**Comment by federico-terzi** [comment]
> yeah, I noticed textexpander actually had a popup warning about process X by name having blocked secure input. would be nice with similar warning.

Yesterday I was fiddling with the idea of a notification to warn the user about it. As far as I could read in the TextExpander docs, it uses a 15 seconds polling to detect SecureInput and that would be quite easy to implement. My only concern is whether it could affect efficiency/battery life.

**Comment by maxandersen** [comment]
turn it of if laptop is running on battery ?

**Comment by federico-terzi** [comment]
> turn it of if laptop is running on battery ?
That could be a valid idea. After a bit of experimentation, I might have found a relatively efficient way to check it which shouldn't have a big impact on battery life. I'll let you know once is ready to try :)
